test/test_testing.py

```diff
@@ -4841,3 +4842,270 @@ def test_errors(self):
+
+        self.harness.handle_exec(
+            self.container, [], result=ExecResult(
+                stdout=b"output", stderr=b"error"))
+        stdout = io.StringIO()
+        stderr = io.StringIO()
+        proc = self.container.exec(["ls"], stderr=stderr, stdout=stdout)
+        self.assertEqual(stdout.getvalue(), "output")
+        self.assertEqual(stderr.getvalue(), "error")
+        proc = self.container.exec(["ls"])
+        self.assertEqual(proc.stdout.read(), "output")
+        self.assertEqual(proc.stderr.read(), "error")
+
+        stdout = io.BytesIO()
+        stderr = io.BytesIO()
+        proc = self.container.exec(["ls"], stderr=stderr, stdout=stdout, encoding=None)
+        self.assertEqual(stdout.getvalue(), b"output")
+        self.assertEqual(stderr.getvalue(), b"error")
+        proc = self.container.exec(["ls"], encoding=None)
+        self.assertEqual(proc.stdout.read(), b"output")
+        self.assertEqual(proc.stderr.read(), b"error")
+
+    def test_exec_service_context(self):
+        service = {
+            "command": "test",
+            "working-dir": "/tmp",
+            "user": "foo",
+            "user-id": 1,
+            "group": "bar",
+            "group-id": 2,
+            "environment": {"foo": "bar", "foobar": "barfoo"}
+        }
+        self.container.add_layer(label="test", layer={
+            "summary": "", "description": "", "services": {"test": service}
+        })
+        args_history = []
+
+        def handler(args):
+            args_history.append(args)
+            return 0
+        os.environ["JUJU_VERSION"] = "3.2.1"
+        self.harness.handle_exec(self.container, ["ls"], handler=handler)
+
+        self.container.exec(["ls"], service_context="test").wait()
+        self.assertEqual(args_history[-1].working_dir, "/tmp")
+        self.assertEqual(args_history[-1].user, "foo")
+        self.assertEqual(args_history[-1].user_id, 1)
+        self.assertEqual(args_history[-1].group, "bar")
+        self.assertEqual(args_history[-1].group_id, 2)
+        self.assertDictEqual(args_history[-1].environment, {"foo": "bar", "foobar": "barfoo"})
+
+        self.container.exec(["ls"],
+                            service_context="test",
+                            working_dir="/test",
+                            user="test",
+                            user_id=3,
+                            group="test_group",
+                            group_id=4,
+                            environment={"foo": "hello"}).wait()
+        self.assertEqual(args_history[-1].working_dir, "/test")
+        self.assertEqual(args_history[-1].user, "test")
+        self.assertEqual(args_history[-1].user_id, 3)
+        self.assertEqual(args_history[-1].group, "test_group")
+        self.assertEqual(args_history[-1].group_id, 4)
+        self.assertDictEqual(args_history[-1].environment, {"foo": "hello", "foobar": "barfoo"})
+
+    def test_registration_order(self):
```

Comment:
What's actually being tested here? That length matters and not registration order? Can we do the same thing without inspecting implementation details, for example just by doing two registrations in one order and then in the opposite order and seeing that the matching is correct?

Reply:
See this reply https://github.com/canonical/operator/pull/993#discussion_r1302970333

Reply:
Let's remove this test once we're using a dict for the registered handlers (or simplify it per my suggestion above in this thread).

Reply:
Removed, thanks!